docs/dg/learningBasics.md

```diff
@@ -0,0 +1,416 @@
+  * `getRepoConfigurations(cliArguments)`: RepoSense gets the configuration for each repository by parsing the CSV files.
+  * `getReportConfigurations(cliArguments)`: RepoSense gets the report configuration (report title) by parsing the JSON files.
+  * `ReportGenerator.generateReposReport(...)` This is where the bulk of the work is conducted, including cloning repositories, analyzing repositories, and generating the JSON files to be used by the report.
+
+<box type="info" seamless>
+
+When tracing through the program execution, you can cross reference the architecture diagram and Javadoc of the class and method to check your understanding of the procedure.
+</box>
+
+{{ step(3) }} Gain some hands-on experience
+
+Here are some small tasks for you to get started on RepoSense backend contribution.
+
+<panel header="**Command Line Argument Parser**" type="primary">
+
+**Command Line Argument Parser**
+
+1. [Generate a report locally](../ug/generatingReports.html#generating-reports-locally).
+2. Open a generated JSON file. For example, you can open `summary.json` of the generated report.
+3. You should see that the content in the JSON is compactly organized in 1 single line. This may affect readability for developer who wants to investigate the content in the JSON file.
+
+**Your Task**
+
+Add a new CLI argument `--use-json-pretty-printing`, such that when a user runs a command `java -jar RepoSense.jar --repos LIST_OF_REPO_URLS --view --use-json-pretty-printing`, the JSON file will be printed in a more readable way.
+
+<panel header="Hint 1">
+
+Try to understand the parsing process of the Command Line Arguments, which starts from `ArgsParser.parse(args)` in `RepoSense.java` and uses `ArgumentParser` to convert an argument string into a Java Object.
+
+Therefore, the first step you can take is to add the following to `ArgsParser`.
+
+```
+public static final String[] JSON_PRINT_MODE_FLAGS = new String[]{"--use-json-pretty-printing", "-j"};
+```
+
+In `getArgumentParser` method, add the following content.
+
+```
+parser.addArgument(JSON_PRINT_MODE_FLAGS)
+    .dest(JSON_PRINT_MODE_FLAGS[0])
+    .action(Arguments.storeTrue())
+    .help("A flag to use json pretty printing when generating the json files.");
+```
+</panel>
+
+<panel header="Hint 2">
+
+After the step in hint 1, the argument is captured by `ArgumentParser`. Now change the  `CliArguments`, `ConfigCliArguments`, `parse` method in `ArgsParser` to make the return result of `parse` include the new argument. 
+
+Add the following content to `CliArguments`.
+
+```
+protected boolean isPrettyPrintingUsed;
+
+public boolean isPrettyPrintingUsed() {
+        return isPrettyPrintingUsed;
+}
+```
+
+In the [constructor](https://github.com/reposense/RepoSense/blob/9125d4b6bea1e5bb6329cb8f7c476ea18fed3cea/src/main/java/reposense/model/ConfigCliArguments.java#L26-L52) of `ConfigCliArguments`, add `isPrettyPrintingUsed` as a new argument of the method, add the following content to the method body.
```

Comment:
Perhaps not fix the link to specific version or specific lines? If you do, it can get out of sync with the actual code sooner or later.

Reply:
Sure, no problem